src/main/java/net/minecraftforge/common/crafting/PartialNBTIngredient.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Minecraft Forge - Forge Development LLC
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.common.crafting;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+import net.minecraft.advancements.critereon.NbtPredicate;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.util.GsonHelper;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.level.ItemLike;
+import net.minecraftforge.registries.ForgeRegistries;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** Ingredient that matches the given items, performing a partial NBT match. Use {@link NBTIngredient} if you want exact match on NBT */
+public class PartialNBTIngredient extends AbstractIngredient
+{
+    private final Set<Item> items;
+    private final CompoundTag nbt;
+    private final NbtPredicate predicate;
+    protected PartialNBTIngredient(Set<Item> items, CompoundTag nbt)
+    {
+        super(items.stream().map(item ->
+            {
+                ItemStack stack = new ItemStack(item);
+                // copy NBT to prevent the stack from modifying the original, as capabilities or vanilla item durability will modify the tag
+                stack.setTag(nbt.copy());
+                return new Ingredient.ItemValue(stack);
+            }));
+        if (items.isEmpty())
+        {
+            throw new IllegalArgumentException("Cannot create a PartialNBTIngredient with no items");
+        }
+        this.items = Collections.unmodifiableSet(items);
+        this.nbt = nbt;
+        this.predicate = new NbtPredicate(nbt);
+    }
+
+    /** Creates a new ingredient matching any item from the list, containing the given NBT */
+    public static PartialNBTIngredient of(CompoundTag nbt, ItemLike... items)
+    {
+        return new PartialNBTIngredient(Arrays.stream(items).map(ItemLike::asItem).collect(Collectors.toSet()), nbt);
```

Comment:
Mostly less redundancy. So less network traffic to sync (sync the tag once instead of once per item) and cleaner recipe JSON. As I said in the main post, it was trivial to do but not strictly needed